File: backend/app/models/task.py
import uuid
from datetime import datetime
from enum import Enum
from sqlalchemy import String, Integer, DateTime, ForeignKey, func, Text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base


class TaskType(str, Enum):
    GENESIS = "genesis"
    MIRROR = "mirror"
    REFINEMENT = "refinement"


class TaskStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(String(50), default=TaskStatus.PENDING.value)
    progress: Mapped[int] = mapped_column(Integer, default=0)
    input_images: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    output_images: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    parameters: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    credits_used: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

File: frontend/src/types/task.ts
// Task types
export type TaskType = 'genesis' | 'mirror' | 'refinement';
export type TaskStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface TaskParameters {
  count?: number;
  style?: string;
  width?: number;
  height?: number;
  referenceImage?: string;
}

export interface Task {
  id: string;
  userId: string;
  type: TaskType;
  status: TaskStatus;
  progress: number;
  inputImages: string[];
  outputImages: string[];
  parameters: TaskParameters;
  createdAt: string;
  error?: string;
}

export interface CreateTaskRequest {
  type: TaskType;
  imageUrl: string;
  count?: number;
  style?: string;
  referenceImage?: string;
  width?: number;
  height?: number;
}

export interface TaskResponse {
  task_id: string;
  status: TaskStatus;
}

export interface TaskProgressUpdate {
  task_id: string;
  progress: number;
  status: TaskStatus;
  output_images?: string[];
  error?: string;
}
